### PR TITLE
Prevent possible leaks

### PR DESCRIPTION
**Pull request by acv**
In some error scenarios, HTTP `request` objects could not have been freed.

Prompted by `valgrind` results from #34

### CODE REVIEW

**Comment by acv** [comment]
3 out of 4 builds in the matrix failed because Travis couldn't `git pull` from github due to ongoing DDoS.

**Comment by acv** [comment]
All build passed now that Github is accessible.

**Comment by mhaas** [comment]
Thanks. I will test tomorrow with the load tester. Now that it's not crashing anymore, I'm seeing some nice memory leaks and (possibly) hanging threads. That is, before applying this patch.

**Comment by mhaas** [comment]
I'm running this patch and your refactor patch and the load tester is behaving a bit weird... it seems that one of the HTTP call raises an exception. When I print it, I only get the empty string ''. Of course, the request does not seem to fire. I will investigate further tomorrow.

File: src/gateway.c [hunk 1]
```diff
@@ -443,7 +443,9 @@ main_loop(void)
          * values that are not -1, 0 or 1. */
         if (webserver->lastError == -1) {
             /* Interrupted system call */
-            continue;           /* restart loop */
+            if (NULL != r) {
```
Comment (mhaas) [inline src/gateway.c]: Shouldn't we have a DEBUG log here? Or is an interrupted system call something that happens regularly?

Comment (acv) [inline src/gateway.c]: It should happen a lot when a gateway is not in use.

Comment (mhaas) [inline src/gateway.c]: No, I don't think so. The timeout argument to select is NULL:

> If timeout is NULL (no time‐out), select() can block indefinitely.

The load tester runs fine until, at some point, I am seeing a lot of bad (empty) HTTP status lines which possibly come from somewhere in that code path. I added a debug print and the bad status lines and webServr->lastError == -1 do coincide.

**Comment by mhaas** [comment]
Ah, I think the problem happens because server->lastError is never reset. This way, the code always takes the error handling path even if the request was actually succesfull.

**Pull request merged**